File: labs/dotnet/lab3-agentic-rag/lab/Agents/AgentFactory.cs
using Lab3.Services;
using Microsoft.Agents.AI;
using OpenAI.Chat;


namespace Lab3.Agents;

/// <summary>
/// Factory for creating specialized agents.
/// </summary>
public class AgentFactory
{
    private readonly ChatClient _chatClient;
    private readonly SearchService _searchService;

    public AgentFactory(ChatClient chatClient, SearchService searchService)
    {
        _chatClient = chatClient;
        _searchService = searchService;
    }

    /// <summary>
    /// Create all agents needed for the system.
    /// </summary>
    public Dictionary<string, AIAgent> CreateAllAgents()
    {
        return new Dictionary<string, AIAgent>
        {
            ["classifier"] = ClassifierAgent.Create(_chatClient),
            ["semantic_search"] = SemanticSearchAgent.Create(_chatClient, _searchService)
            // TODO: Add other agents here as needed
        };
    }
}

File: labs/python/lab3-agentic-rag/begin/agents/agent_factory.py
"""
Agent factory for creating all specialized agents.

This module imports individual agent creation functions and provides
a unified AgentFactory class for easy agent instantiation.
"""
from agent_framework import ChatAgent
from agent_framework.azure import AzureOpenAIChatClient

from agents import (
    classifier_agent,
    semantic_search_agent,
)
from services import SearchService

class AgentFactory:
    """Factory for creating specialized agents."""
    
    def __init__(self, chat_client: AzureOpenAIChatClient, search_service: SearchService):
        """
        Initialize the agent factory.
        
        Args:
            chat_client: Azure OpenAI chat client
            search_service: Search service for ticket queries
        """
        self.chat_client = chat_client
        self.search_service = search_service
    
    def create_all_agents(self) -> dict[str, ChatAgent]:
        """
        Create all agents needed for the system.
        
        Returns:
            Dictionary mapping agent names to ChatAgent instances
        """
        return {
            "classifier": classifier_agent.create_classifier_agent(self.chat_client),
            "semantic_search": semantic_search_agent.create_semantic_search_agent(self.chat_client, self.search_service),
            # TODO: Add more agents here as needed
        }
